kratos/utilities/math_utils.h

```diff
@@ -738,37 +676,77 @@ class MathUtils
      * @param rA Is the input matrix
      * @return The determinant of the 4x4 matrix
      */
-    static inline TDataType Det4(const MatrixType& rA)
+    template<class TMatrixType>
+    static inline TDataType Det4(const TMatrixType& rA)
     {
-        const double Det = rA(0,1)*rA(1,3)*rA(2,2)*rA(3,0)-rA(0,1)*rA(1,2)*rA(2,3)*rA(3,0)-rA(0,0)*rA(1,3)*rA(2,2)*rA(3,1)+rA(0,0)*rA(1,2)*rA(2,3)*rA(3,1)
+        const double det = rA(0,1)*rA(1,3)*rA(2,2)*rA(3,0)-rA(0,1)*rA(1,2)*rA(2,3)*rA(3,0)-rA(0,0)*rA(1,3)*rA(2,2)*rA(3,1)+rA(0,0)*rA(1,2)*rA(2,3)*rA(3,1)
                           -rA(0,1)*rA(1,3)*rA(2,0)*rA(3,2)+rA(0,0)*rA(1,3)*rA(2,1)*rA(3,2)+rA(0,1)*rA(1,0)*rA(2,3)*rA(3,2)-rA(0,0)*rA(1,1)*rA(2,3)*rA(3,2)+rA(0,3)*(rA(1,2)*rA(2,1)*rA(3,0)-rA(1,1)*rA(2,2)*rA(3,0)-rA(1,2)*rA(2,0)*rA(3,1)+rA(1,0)*rA(2,2)*rA(3,1)+rA(1,1)*rA(2,0)*rA(3,2)
                           -rA(1,0)*rA(2,1)*rA(3,2))+(rA(0,1)*rA(1,2)*rA(2,0)-rA(0,0)*rA(1,2)*rA(2,1)-rA(0,1)*rA(1,0)*rA(2,2)+rA(0,0)*rA(1,1)*rA(2,2))*rA(3,3)+rA(0,2)*(-(rA(1,3)*rA(2,1)*rA(3,0))+rA(1,1)*rA(2,3)*rA(3,0)+rA(1,3)*rA(2,0)*rA(3,1)-rA(1,0)*rA(2,3)*rA(3,1)-rA(1,1)*rA(2,0)*rA(3,3)+rA(1,0)*rA(2,1)*rA(3,3));
-        return Det;
+        return det;
     }
 
     /**
-     * @brief Calculates the determinant of a matrix of dimension 2x2 (in this case for a bounded matrix)
+     * @brief Calculates the determinant of a matrix of dimension 2x2 or 3x3 (no check performed on matrix size)
      * @param rA Is the input matrix
-     * @return The determinant of the matrix
+     * @return The determinant of the 2x2 matrix
      */
-    static inline TDataType Det(const BoundedMatrix<double,2,2>& rA)
+    template<class TMatrixType>
+    static inline TDataType Det(const TMatrixType& rA)
     {
-        return (rA(0,0)*rA(1,1)-rA(0,1)*rA(1,0));
+        switch (rA.size1()) {
```

Comment:
Just wondering if we should check in here that the matrix is square.

Reply:
I will not, as it has a penalty, maybe in debug. In any case if you want to compute the determiannt a non square you should use the GeneralizedDet method instead

Reply:
Makes sense. I think we can do it only in debug mode.

Reply:
Done